### ISSUE TITLE
Running my script freezes the website

### ISSUE THREAD

**Issue by QuentinMcCarthy** [bitburner-official/bitburner-src#1082]
I've stored servers from a full scan server in a .txt file using a script with details such as the minimum security level, maximum money and amount of ports needed to be open. I'm then accessing the details from the .txt file in another script to use them for the purposes of hacking the servers.

Via debugging, I'm able to access the full server list object from the .txt file, and able to access all individual values using both dot notation and brace notation. However, as soon as I try to access the server's required ports value in an if statement within a for loop, it claims the value is undefined.

I've been debugging this for a bit now and it's starting to get frustrating.

The object (only showing relevant entry) is:
```JSON
{
  "omega-net":{
    "maxMon":1748919275,
    "minSec":10,
    "ports":2
  }
}
```
While the offending code is:
```javascript
for (let i = 0; serverList.length; i++) {
    let debug = "ServerList: "+JSON.stringify(serverList)+"\n";
    debug += "Servers: "+JSON.stringify(servers)+"\n";
    debug += "CurrServer: "+JSON.stringify(servers[serverList[i]])+"\n";
    debug += "Ports: "+servers[serverList[i]]["ports"]+"\n";
    debug += "Ports again: "+servers[serverList[i]].ports;

    ns.write("data/debug.txt",debug,"w");
    
    // Can we open enough ports?
    if (servers[serverList[i]].ports <= canOpen) {
      if (canSSH) { ns.brutessh(serverList[i]); }
      if (canFTP) { ns.ftpcrack(serverList[i]); }
      if (canSMTP) { ns.relaysmtp(serverList[i]); }
      if (canHTTP) { ns.httpworm(serverList[i]); }

      // Get root on the server
      ns.nuke(serverList[i]);
    } else {
      // Remove servers from the list if we can't access them
      serverList.splice(i,1);
    }
  }
```
The debug stuff exports to a text file which reports the correct values (again only showing relevant):
```
CurrServer: {"maxMon":1748919275,"minSec":10,"ports":2}
Ports: 2
Ports again: 2
```
However trying to run the script reports an error at the `if (servers[serverList[i]].ports <= canOpen) {` line, claiming "cannot read properties of undefined".

The "servers" variable is defined at the top of the script as a const, defined as what's imported from the .txt file. It is always defined and correct.
The "serverList" variable is defined immediately below the "servers" variable at the top of the script, as a var, defined as all the keys in the "servers" object. It is always defined and correct.

I've attached the save file which contains all relevant files found in home as scripts/hack-all.js and data/servers.txt
[bitburnerSave_debug.json](https://github.com/bitburner-official/bitburner-src/files/14218562/bitburnerSave_debug.json)

**Comment by QuentinMcCarthy** [comment]
I just attempted to put the ports into its own variable to access it that way, however it won't even let me do that, claiming it's undefined when trying to assign it to the variable...

**Comment by QuentinMcCarthy** [comment]
I figured it out. For some reason the length of the serverList array is 15 with the 15th entry being undefined. The undefined entry doesn't appear when I print the array in its entirety to the debug.txt file, but if I iterate and append to the debug.txt in the for loop, it reports all entries plus a 15th entry of "undefined". Simply adding an if statement to ignore the undefined entry + remove it from the array fixes the issue.

**Comment by QuentinMcCarthy** [comment]
Ok now I have to reopen this, I'm facing another issue.
If I run the fixed script, the game freezes and locks up. It won't continue until I close the tab and reopen the site.

New save with the fixed script causing the issue:
[bitburnerSave_debug.json](https://github.com/bitburner-official/bitburner-src/files/14219573/bitburnerSave_debug.json)

Only thing I can think is it's being caused by the while(true) loop, but I don't know why because it's the exact same loop used in the documentation/tutorial, and I even have the original version of this script with the exact same loop that still works.

Edit:
Just tried a different method, putting all the details into three arrays inside an object instead of all as object values... Still froze. This narrows down my thinking to the issue being something to do with objects.

Edit2:
Narrowing down the thinking more, I think it's related to the importing of the data from the text file. I set everything as arrays and it's still freezing

Edit3:
Nope, just tried to put the data directly into the script instead of importing from the text file, still freezes....

**Comment by d0sboots** [comment]
It is common to get freezes if you have created an infinite loop. There are a variety of ways to debug this; you can use Chrome's devtools (accessible with Debug > Activate in the Steam version), or adding prints + sleeps, or other methods.

If you run into a situation where you have a script that freezes as soon soon as the game starts, you can run the game without any scripts loaded (there is an option for this in steam, or it's a URL param in the web version.)

**Comment by QuentinMcCarthy** [comment]
I figured as much regarding infinite loops, however I'm certain that's not the cause. I'm currently actively using the exact same code as in the above saves (only slightly altered as I've iterated on it a bit to improve it), with the only major difference being I'm manually entering the server names into an array instead of trying to pull it from the .txt file. Everything else is the same. Same loops, same if statements, same calls...

So it's definitely not an infinite loop issue, I can assure you. As for accessing the dev tools, when the site freezes I'm no longer able to open the dev tools - I've tried - I'm completely locked out of the tab, with the only functionality that allows me to interact with the tab being to, well, close it (not counting moving the tab around). I can't even change the url in the tab, it completely locks up. Additionally, trying to open the devtools BEFORE running the script produces no errors. I'm uncertain if it's simply not displaying due to the tab locking up, but...

As an addendum, I use Opera GX, the entire rest of the browser is completely unaffected by bitburner locking up (I watch YouTube while playing on my second monitor, I'd notice if it locked up fully)

Oh, and one final thing. If it was an infinite loop issue, it shouldn't affect the script when there are no loops, yet it does. All I'm doing is accessing a variable, shouldn't be any loops to cause issues with.

Edit:
Should mention, as demonstrated in the original message for this thread when I was having a different issue, one of my first methods of debugging is printing (I only after making the original script discovered ns.print, was trying to use console.log which did nothing in the tail window, so ended up using a debug.txt file) but there are no errors to be found in my logging. None at all. Bear in mind that because I was logging to a debug.txt file instead of the tail window, even if I restarted the tab, the log should still be in the debug.txt as it would've written before locking up, but everything in the debug log is fine.

**Comment by Caldwell-74** [comment]
when it freezes and you reload it loads the last save 
this save is guaranteed to be before your freeze so your logging to a file wont be included
console.log prints to the dev console even when the context is frozen
why not just use a for of loop?

**Comment by Caldwell-74** [comment]
yeah found the bug
`for (let i = 0; serverList.length; i++) {`
this doesnt work because the end condition is that serverList is empty
if there is even a single server you can nuke it wont be removed keeping the loop running forever
at some point i will be bigger than the serverList.length and then you just access undefined forever
your undefined check doesnt prevent that at all

btw for script debugging come to the discord thats a better place for that kind of stuff
https://discord.gg/AafjRUNN

```
/** @param {NS} ns */
export async function main(ns) {
  // Pull server details created by get-servers.js
  const servers = JSON.parse(ns.read("data/servers.txt"));
  const serverList = Object.keys(servers);

  // What ports can we open?
  let canOpen = 0;
  let canSSH = false;
  let canFTP = false;
  let canSMTP = false;
  let canHTTP = false;
  if (ns.fileExists("BruteSSH.exe", "home")) {
    canSSH = true;
    canOpen++;
  }
  if (ns.fileExists("FTPCrack.exe", "home")) {
    canFTP = true;
    canOpen++;
  }
  if (ns.fileExists("RelaySMTP.exe", "home")) {
    canSMTP = true;
    canOpen++;
  }
  if (ns.fileExists("HTTPWorm.exe", "home")) {
    canHTTP = true;
    canOpen++;
  }

  for (const server of serverList) {
    // Can we open enough ports?
    if (servers[server].ports <= canOpen) {
      if (canSSH) { ns.brutessh(server); }
      if (canFTP) { ns.ftpcrack(server) }
      if (canSMTP) { ns.relaysmtp(server); }
      if (canHTTP) { ns.httpworm(server); }

      // Get root on the server
      ns.nuke(server);
    }
  }
  const nukedServers = serverList.filter(server => ns.hasRootAccess(server))

  // Continuously hack/grow/weaken the servers
  while (true) {
    for (const server of nukedServers) {
      const currServer = servers[server];

      if (ns.getServerSecurityLevel(server) > currServer["minSec"]) {
        await ns.weaken(server);
      } else if (ns.getServerMoneyAvailable(server) < currServer["maxMon"]) {
        await ns.grow(server);
      } else {
        await ns.hack(server);
      }
    }
  }
}
```

**Comment by Snarling** [comment]
Closing issue as this is a bug in a player script and not a bug in the game (for loop condition creates infinite loop with no delays)

**Comment by QuentinMcCarthy** [comment]
Ah, I see, I apologise. I misread the description of the Object.keys() function, thinking it output an array, not an object.
